tests/test_bqpipeline.py

```diff
@@ -63,24 +77,116 @@ def test_create_job_config_destination(self):
             self.assertEqual(cfg.destination.project, 'testproject')
 
     def test_create_job_config_flags(self):
-        bq = BQPipeline(
-            job_name='testjob', default_project='testproject', default_dataset='testdataset')
+        bq = bqpipeline.BQPipeline(
+            job_name='testjob', default_project='testproject',
+            default_dataset='testdataset')
         cfg = bq.create_job_config(batch=False, create=False, overwrite=False)
         self.assertEqual(cfg.priority, bigquery.QueryPriority.INTERACTIVE)
         self.assertEqual(cfg.create_disposition,
                          bigquery.job.CreateDisposition.CREATE_NEVER)
         self.assertEqual(cfg.write_disposition,
                          bigquery.job.WriteDisposition.WRITE_EMPTY)
 
+    def test_scalar_parameters(self):
+        result = bqpipeline.set_scalar_parameter('test', 'abc', typ='str')
+        self.assertEqual(result, bigquery.ScalarQueryParameter('test',
+                                                               'STRING', 'abc'))
+
+        result = bqpipeline.set_scalar_parameter('test', 1, typ='int')
+        self.assertEqual(result, bigquery.ScalarQueryParameter('test',
+                                                               'INT64', 1))
+
+        dt = datetime.datetime.now()
+        result = bqpipeline.set_scalar_parameter('test', dt, typ='datetime')
+        self.assertEqual(result, bigquery.ScalarQueryParameter('test',
+                                                               'TIMESTAMP', dt))
+
+        b = bytes(123)
+        result = bqpipeline.set_scalar_parameter('test', b, typ='bytes')
+        self.assertEqual(result, bigquery.ScalarQueryParameter('test',
+                                                               'BYTES', b))
+
+        with self.assertRaises(KeyError):
+            bqpipeline.set_scalar_parameter('test', 1, typ='unsupported')
+
+    def test_array_parameters(self):
+        result = bqpipeline.set_list_parameter('test', ['abc'], typ='list')
+        self.assertEqual(result, bigquery.ArrayQueryParameter('test', 'STRING',
+                                                              ['abc']))
+
+        result = bqpipeline.set_list_parameter('test', [1, 2], typ='list')
+        self.assertEqual(result, bigquery.ArrayQueryParameter('test', 'INT64',
+                                                              [1, 2]))
+
+        with self.assertRaises(IndexError):
+            bqpipeline.set_list_parameter('test', [], typ='list')
+
+    def test_struct_parameters(self):
+        result = bqpipeline.set_dict_parameter('test', {'a': 'abc'},
+                                               typ='list')
+        self.assertEqual(result, bigquery.StructQueryParameter(
```

Comment:
Add test for empty dict. Should we error on empty dict?

Reply:
For an empty dict or a list, the use case I was thinking about was when the query is trying to set the default value to empty dict or list respectively. In case of list, it is not possible to set such parameters without the knowledge of the array type - we currently infer that from the contents of the python list. In case of a dict, however, that problem does not exist as the struct can be set with empty contents without having to infer any type info for the contents. Let me know your thoughts on the same.

Reply:
I disagree about dict.
A struct w/o any keys makes no in BQ. A struct can either be `null` or a struct w/ keys whose values are null. 
I'd be surprised if BQ didn't throw an error for passing a query param of `{}`

Reply:
Done.